File: C/Claude/NBodySimulation.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#define G 6.67430e-11  // Gravitational constant
#define SOLAR_MASS 1.989e30  // Mass of the Sun in kg
#define AU 1.496e11  // Astronomical Unit in meters
#define YEAR 3.154e7  // Year in seconds

typedef struct {
    double x, y, z;       // Position
    double vx, vy, vz;    // Velocity
    double mass;          // Mass
} Body;

// Vector operations
double vector_distance(double x1, double y1, double z1, double x2, double y2, double z2) {
    double dx = x2 - x1;
    double dy = y2 - y1;
    double dz = z2 - z1;
    return sqrt(dx*dx + dy*dy + dz*dz);
}

// Calculate total energy of the system (kinetic + potential)
double calculate_energy(Body *bodies, int n) {
    double energy = 0.0;
    
    // Calculate kinetic and potential energy
    for (int i = 0; i < n; i++) {
        // Kinetic energy
        double v_squared = bodies[i].vx * bodies[i].vx + 
                           bodies[i].vy * bodies[i].vy + 
                           bodies[i].vz * bodies[i].vz;
        energy += 0.5 * bodies[i].mass * v_squared;
        
        // Potential energy with all other bodies
        for (int j = i + 1; j < n; j++) {
            double distance = vector_distance(
                bodies[i].x, bodies[i].y, bodies[i].z,
                bodies[j].x, bodies[j].y, bodies[j].z
            );
            energy -= G * bodies[i].mass * bodies[j].mass / distance;
        }
    }
    
    return energy;
}

// Update velocities of all bodies
void update_velocities(Body *bodies, int n, double dt) {
    for (int i = 0; i < n; i++) {
        double fx = 0.0, fy = 0.0, fz = 0.0;
        
        for (int j = 0; j < n; j++) {
            if (i == j) continue;
            
            double dx = bodies[j].x - bodies[i].x;
            double dy = bodies[j].y - bodies[i].y;
            double dz = bodies[j].z - bodies[i].z;
            
            double distance = sqrt(dx*dx + dy*dy + dz*dz);
            double distance_cubed = distance * distance * distance;
            
            // Skip if bodies are too close (to avoid numerical instability)
            if (distance < 1e-10) continue;
            
            // Calculate gravitational force
            double force = G * bodies[i].mass * bodies[j].mass / distance_cubed;
            
            fx += force * dx;
            fy += force * dy;
            fz += force * dz;
        }
        
        // Update velocities (F = ma => a = F/m)
        bodies[i].vx += fx / bodies[i].mass * dt;
        bodies[i].vy += fy / bodies[i].mass * dt;
        bodies[i].vz += fz / bodies[i].mass * dt;
    }
}

// Update positions of all bodies
void update_positions(Body *bodies, int n, double dt) {
    for (int i = 0; i < n; i++) {
        bodies[i].x += bodies[i].vx * dt;
        bodies[i].y += bodies[i].vy * dt;
        bodies[i].z += bodies[i].vz * dt;
    }
}

// Kick-step integration (first-order symplectic)
void simulate_step(Body *bodies, int n, double dt) {
    update_velocities(bodies, n, dt);  // Kick
    update_positions(bodies, n, dt);   // Drift
}

// Initialize system with central body and orbiting bodies
void initialize_system(Body *bodies, int n_bodies, double central_mass, double orbit_radius_min, double orbit_radius_max) {
    if (n_bodies < 1) return;
    
    // Set central body at origin
    bodies[0].x = 0.0;
    bodies[0].y = 0.0;
    bodies[0].z = 0.0;
    bodies[0].vx = 0.0;
    bodies[0].vy = 0.0;
    bodies[0].vz = 0.0;
    bodies[0].mass = central_mass;
    
    srand(time(NULL));
    
    // Create orbiting bodies
    for (int i = 1; i < n_bodies; i++) {
        // Random angle for position
        double phi = ((double)rand() / RAND_MAX) * 2.0 * M_PI;
        double theta = ((double)rand() / RAND_MAX) * M_PI - M_PI/2.0;
        
        // Random distance within range
        double radius = orbit_radius_min + ((double)rand() / RAND_MAX) * (orbit_radius_max - orbit_radius_min);
        
        // Position on sphere
        bodies[i].x = radius * cos(theta) * cos(phi);
        bodies[i].y = radius * cos(theta) * sin(phi);
        bodies[i].z = radius * sin(theta);
        
        // Velocity for circular orbit
        double orbit_speed = sqrt(G * central_mass / radius);
        
        // Cross product with up vector to get perpendicular velocity
        bodies[i].vx = -orbit_speed * sin(phi);
        bodies[i].vy = orbit_speed * cos(phi);
        bodies[i].vz = 0.0;
        
        // Small mass compared to central body
        bodies[i].mass = central_mass * 1e-6;
    }
}

void print_system_stats(Body *bodies, int n) {
    printf("System statistics:\n");
    printf("Central body: position (%.2e, %.2e, %.2e), mass: %.2e\n", 
           bodies[0].x, bodies[0].y, bodies[0].z, bodies[0].mass);
    
    // Calculate system boundaries
    double min_x = bodies[0].x, max_x = bodies[0].x;
    double min_y = bodies[0].y, max_y = bodies[0].y;
    double min_z = bodies[0].z, max_z = bodies[0].z;
    
    for (int i = 1; i < n; i++) {
        if (bodies[i].x < min_x) min_x = bodies[i].x;
        if (bodies[i].x > max_x) max_x = bodies[i].x;
        if (bodies[i].y < min_y) min_y = bodies[i].y;
        if (bodies[i].y > max_y) max_y = bodies[i].y;
        if (bodies[i].z < min_z) min_z = bodies[i].z;
        if (bodies[i].z > max_z) max_z = bodies[i].z;
    }
    
    printf("System boundaries:\n");
    printf("  X: [%.2e, %.2e]\n", min_x, max_x);
    printf("  Y: [%.2e, %.2e]\n", min_y, max_y);
    printf("  Z: [%.2e, %.2e]\n", min_z, max_z);
}

int main() {
    // Problem parameters
    const int N = 10000 + 1;  // 1 central + 1 million orbiting bodies
    const int STEPS = 100;
    const double dt = 0.01 * YEAR;  // Timestep of 0.01 years
    
    printf("Starting N-body simulation with %d bodies for %d steps\n", N, STEPS);
    printf("Allocating memory...\n");
    
    // Allocate memory for bodies
    Body *bodies = (Body *)malloc(N * sizeof(Body));
    if (!bodies) {
        printf("Memory allocation failed!\n");
        return 1;
    }
    
    printf("Initializing system...\n");
    
    // Initialize system with Sun-like central body and Earth-like orbits
    initialize_system(bodies, N, SOLAR_MASS, 0.8 * AU, 5.0 * AU);
    
    print_system_stats(bodies, N);
    
    // Calculate initial energy
    printf("Calculating initial energy...\n");
    double initial_energy = calculate_energy(bodies, N);
    printf("Initial energy: %.10e\n", initial_energy);
    
    printf("Starting simulation...\n");
    clock_t start = clock();
    
    // Run simulation
    for (int step = 0; step < STEPS; step++) {
        if (step % 100 == 0) {
            printf("Step %d of %d (%.1f%%)\n", step, STEPS, 100.0 * step / STEPS);
        }
        simulate_step(bodies, N, dt);
    }
    
    clock_t end = clock();
    double cpu_time = (double)(end - start) / CLOCKS_PER_SEC;
    printf("Simulation completed in %.2f seconds\n", cpu_time);
    
    // Calculate final energy
    printf("Calculating final energy...\n");
    double final_energy = calculate_energy(bodies, N);
    printf("Final energy: %.10e\n", final_energy);
    
    // Calculate energy error
    double energy_error = (final_energy - initial_energy) / initial_energy;
    printf("Relative energy error: %.10e (%.10f%%)\n", energy_error, energy_error * 100.0);
    
    print_system_stats(bodies, N);
    
    // Free memory
    free(bodies);
    
    return 0;
}

File: Java/Claude/NBodySimulation.java
import java.util.Random;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class NBodySimulation {
    // Gravitational constant (using an arbitrary value for simulation purposes)
    private static final double G = 6.67430e-11;
    
    // Simulation bodies
    private Body[] bodies;
    
    // Simulation parameters
    private double timeStep;
    private int numSteps;
    
    // Class to represent a body in the simulation
    static class Body {
        // Position components
        double x, y, z;
        // Velocity components
        double vx, vy, vz;
        // Mass
        double mass;
        
        public Body(double x, double y, double z, double vx, double vy, double vz, double mass) {
            this.x = x;
            this.y = y;
            this.z = z;
            this.vx = vx;
            this.vy = vy;
            this.vz = vz;
            this.mass = mass;
        }
        
        // Copy constructor
        public Body(Body other) {
            this.x = other.x;
            this.y = other.y;
            this.z = other.z;
            this.vx = other.vx;
            this.vy = other.vy;
            this.vz = other.vz;
            this.mass = other.mass;
        }
    }
    
    public NBodySimulation(int numBodies, double timeStep, int numSteps) {
        this.bodies = new Body[numBodies];
        this.timeStep = timeStep;
        this.numSteps = numSteps;
    }
    
    /**
     * Initialize a central body with specified number of smaller bodies on circular orbits
     * @param centralMass Mass of the central body
     * @param bodyMass Mass of each orbiting body
     * @param minRadius Minimum orbit radius
     * @param maxRadius Maximum orbit radius
     */
    public void initializeSystem(double centralMass, double bodyMass, double minRadius, double maxRadius) {
        // Central body at origin with no initial velocity
        bodies[0] = new Body(0, 0, 0, 0, 0, 0, centralMass);
        
        // Create smaller bodies on circular orbits
        Random random = new Random(42); // Use fixed seed for reproducibility
        
        for (int i = 1; i < bodies.length; i++) {
            // Random radius between minRadius and maxRadius
            double radius = minRadius + random.nextDouble() * (maxRadius - minRadius);
            
            // Random position on a sphere of the given radius
            double theta = random.nextDouble() * 2 * Math.PI;
            double phi = Math.acos(2 * random.nextDouble() - 1);
            
            double x = radius * Math.sin(phi) * Math.cos(theta);
            double y = radius * Math.sin(phi) * Math.sin(theta);
            double z = radius * Math.cos(phi);
            
            // Calculate velocity for a circular orbit
            // v = sqrt(G * M / r)
            double velocity = Math.sqrt(G * centralMass / radius);
            
            // Direction perpendicular to radius vector (for circular orbit)
            // We need a vector perpendicular to (x, y, z)
            // One approach is to cross product with a reference vector
            double[] perpDirection = crossProduct(x, y, z, 0, 0, 1);
            
            // If the result is close to zero, use a different reference vector
            if (magnitude(perpDirection) < 1e-10) {
                perpDirection = crossProduct(x, y, z, 0, 1, 0);
            }
            
            // Normalize the perpendicular direction
            double perpMagnitude = magnitude(perpDirection);
            perpDirection[0] /= perpMagnitude;
            perpDirection[1] /= perpMagnitude;
            perpDirection[2] /= perpMagnitude;
            
            // Set velocity components
            double vx = velocity * perpDirection[0];
            double vy = velocity * perpDirection[1];
            double vz = velocity * perpDirection[2];
            
            bodies[i] = new Body(x, y, z, vx, vy, vz, bodyMass);
        }
    }
    
    // Helper function to compute cross product
    private double[] crossProduct(double x1, double y1, double z1, double x2, double y2, double z2) {
        double[] result = new double[3];
        result[0] = y1 * z2 - z1 * y2;
        result[1] = z1 * x2 - x1 * z2;
        result[2] = x1 * y2 - y1 * x2;
        return result;
    }
    
    // Helper function to compute magnitude of a vector
    private double magnitude(double[] vector) {
        return Math.sqrt(vector[0] * vector[0] + vector[1] * vector[1] + vector[2] * vector[2]);
    }
    
    // Calculate total energy (kinetic + potential) of the system
    public double calculateTotalEnergy() {
        double totalEnergy = 0.0;
        
        // Calculate kinetic energy for each body
        for (Body body : bodies) {
            double speedSquared = body.vx * body.vx + body.vy * body.vy + body.vz * body.vz;
            double kineticEnergy = 0.5 * body.mass * speedSquared;
            totalEnergy += kineticEnergy;
        }
        
        // Calculate potential energy for each pair of bodies
        for (int i = 0; i < bodies.length; i++) {
            for (int j = i + 1; j < bodies.length; j++) {
                Body bi = bodies[i];
                Body bj = bodies[j];
                
                double dx = bi.x - bj.x;
                double dy = bi.y - bj.y;
                double dz = bi.z - bj.z;
                
                double distance = Math.sqrt(dx * dx + dy * dy + dz * dz);
                
                // Avoid division by zero
                if (distance > 0) {
                    double potentialEnergy = -G * bi.mass * bj.mass / distance;
                    totalEnergy += potentialEnergy;
                }
            }
        }
        
        return totalEnergy;
    }
    
    // Run the simulation using a first-order kick-step method (leapfrog)
    public void runSimulation() {
        // Display initial energy
        double initialEnergy = calculateTotalEnergy();
        System.out.println("Initial total energy: " + initialEnergy);
        
        // Run simulation steps
        for (int step = 0; step < numSteps; step++) {
            if (step % 100 == 0) {
                System.out.println("Step " + step + " of " + numSteps);
            }
            
            updateVelocities();
            updatePositions();
        }
        
        // Display final energy
        double finalEnergy = calculateTotalEnergy();
        System.out.println("Final total energy: " + finalEnergy);
        System.out.println("Energy change: " + (finalEnergy - initialEnergy));
        System.out.println("Relative energy change: " + ((finalEnergy - initialEnergy) / initialEnergy));
    }
    
    // Update velocities using forces (kick step)
    private void updateVelocities() {
        // For very large systems, use parallel processing
        if (bodies.length > 10000) {
            updateVelocitiesParallel();
            return;
        }
        
        // Calculate forces and update velocities
        for (int i = 0; i < bodies.length; i++) {
            Body bi = bodies[i];
            double fx = 0, fy = 0, fz = 0;
            
            // Calculate net force on body i from all other bodies
            for (int j = 0; j < bodies.length; j++) {
                if (i == j) continue;
                
                Body bj = bodies[j];
                
                double dx = bj.x - bi.x;
                double dy = bj.y - bi.y;
                double dz = bj.z - bi.z;
                
                double distanceSquared = dx * dx + dy * dy + dz * dz;
                double distance = Math.sqrt(distanceSquared);
                
                // Add small softening parameter to avoid numerical instability
                double softening = 1e-10;
                distanceSquared += softening;
                
                // Calculate gravitational force magnitude (F = G * m1 * m2 / r^2)
                double forceMagnitude = G * bi.mass * bj.mass / distanceSquared;
                
                // Force components (F * direction)
                fx += forceMagnitude * dx / distance;
                fy += forceMagnitude * dy / distance;
                fz += forceMagnitude * dz / distance;
            }
            
            // Update velocity using F = ma => a = F/m => Δv = a * Δt
            bi.vx += fx / bi.mass * timeStep;
            bi.vy += fy / bi.mass * timeStep;
            bi.vz += fz / bi.mass * timeStep;
        }
    }
    
    // Update velocities in parallel for large systems
    private void updateVelocitiesParallel() {
        int processors = Runtime.getRuntime().availableProcessors();
        ExecutorService executor = Executors.newFixedThreadPool(processors);
        
        // Calculate forces and update velocities in parallel
        int batchSize = Math.max(1, bodies.length / (processors * 2));
        
        for (int start = 0; start < bodies.length; start += batchSize) {
            final int s = start;
            final int end = Math.min(s + batchSize, bodies.length);
            
            executor.execute(() -> {
                for (int i = s; i < end; i++) {
                    Body bi = bodies[i];
                    double fx = 0, fy = 0, fz = 0;
                    
                    // Calculate net force on body i from all other bodies
                    for (int j = 0; j < bodies.length; j++) {
                        if (i == j) continue;
                        
                        Body bj = bodies[j];
                        
                        double dx = bj.x - bi.x;
                        double dy = bj.y - bi.y;
                        double dz = bj.z - bi.z;
                        
                        double distanceSquared = dx * dx + dy * dy + dz * dz;
                        double distance = Math.sqrt(distanceSquared);
                        
                        // Add small softening parameter to avoid numerical instability
                        double softening = 1e-10;
                        distanceSquared += softening;
                        
                        // Calculate gravitational force magnitude (F = G * m1 * m2 / r^2)
                        double forceMagnitude = G * bi.mass * bj.mass / distanceSquared;
                        
                        // Force components (F * direction)
                        fx += forceMagnitude * dx / distance;
                        fy += forceMagnitude * dy / distance;
                        fz += forceMagnitude * dz / distance;
                    }
                    
                    // Update velocity using F = ma => a = F/m => Δv = a * Δt
                    bi.vx += fx / bi.mass * timeStep;
                    bi.vy += fy / bi.mass * timeStep;
                    bi.vz += fz / bi.mass * timeStep;
                }
            });
        }
        
        executor.shutdown();
        try {
            executor.awaitTermination(1, TimeUnit.HOURS);
        } catch (InterruptedException e) {
            System.err.println("Parallel execution interrupted: " + e.getMessage());
        }
    }
    
    // Update positions using velocities (drift step)
    private void updatePositions() {
        for (Body body : bodies) {
            body.x += body.vx * timeStep;
            body.y += body.vy * timeStep;
            body.z += body.vz * timeStep;
        }
    }
    
    // Main method to run the simulation
    public static void main(String[] args) {
        // Number of bodies (1 central + orbiting bodies)
        final int NUM_BODIES = 10_001; // 1 million small bodies + 1 central body
        
        // Time step (smaller is more accurate but slower)
        final double TIME_STEP = 0.001 * 3600*24*365;
        
        // Number of simulation steps
        final int NUM_STEPS = 100;
        
        // Create simulation
        NBodySimulation simulation = new NBodySimulation(NUM_BODIES, TIME_STEP, NUM_STEPS);
        
        // Central mass (e.g., like a star)
        double centralMass = 1.0e30;
        
        // Orbiting body mass (e.g., like planets)
        double bodyMass = 1.0e20;
        
        // Orbit radius range
        double minRadius = 1.0e8;
        double maxRadius = 5.0e8;
        
        System.out.println("Initializing " + (NUM_BODIES - 1) + " bodies around central mass...");
        long startTime = System.currentTimeMillis();
        
        // Initialize the system
        simulation.initializeSystem(centralMass, bodyMass, minRadius, maxRadius);
        
        long initTime = System.currentTimeMillis() - startTime;
        System.out.println("Initialization completed in " + initTime + " ms");
        
        // Run the simulation
        System.out.println("Running simulation for " + NUM_STEPS + " steps...");
        startTime = System.currentTimeMillis();
        
        simulation.runSimulation();
        
        long simTime = System.currentTimeMillis() - startTime;
        System.out.println("Simulation completed in " + simTime + " ms");
    }
}